NSI utils for obtaining NSI data

ywkim312 opened the pull request:
please test it by running test_nsi.py in tests/pyincore_data folder

navarroc commented:
Should nsiutil be inside the "utils" folder?

navarroc commented:
I see censusutil is also at the top level, which is kind of odd since there is a utils folder. At some point we probably need to think about the structure of pyincore-data. Something like:
pyincore_data / utils / nsiutil
pyincore_data / nsi / nsiparser.py (which calls nsiutil maybe)

or it could be pyincore_data/nsiparser.py

ywkim312 commented:
> Should nsiutil be inside the "utils" folder?

That is a good question. I had a hard time naming it, and there was a censusutil in the same place, so I named it that way. Do you have any other idea for the name of this nsiutil? I think that this file is better positioned in its current location, but maybe the name should be changed. If you propose a different name, I will update the code accordingly.

ywkim312 commented:
> I see censusutil is also at the top level, which is kind of odd since there is a utils folder. At some point we probably need to think about the structure of pyincore-data. Something like: pyincore_data / utils / nsiutil pyincore_data / nsi / nsiparser.py (which calls nsiutil maybe)
> 
> or it could be pyincore_data/nsiparser.py

NSI parser sounds good. I will update using this name. Do you have other idea for censusutil as well?

longshuicy commented on pyincore_data/utils/datautil.py:
```diff
@@ -51,3 +62,177 @@ def convert_dislocation_pd_to_csv(in_pd, save_columns, programname, savefile):
+        file_name = cfg.NSI_PREFIX + str(state_fips) + ".gpkg.zip"
+        file_url = "%s/%s" % (cfg.NSI_URL_STATE, file_name)
+        print("Downloading NSI data for the state: " + str(state_fips))
+        r = requests.get(file_url, stream=True)
```
Just curious, no need for `result.raise_for_status()` here?

ywkim312 replied:
added

longshuicy commented on pyincore_data/utils/datautil.py:
```diff
@@ -51,3 +62,177 @@ def convert_dislocation_pd_to_csv(in_pd, save_columns, programname, savefile):
+        with open(download_filename, "wb") as zipfile:
+            for chunk in r.iter_content(chunk_size=1024):
+                if chunk:
+                    zipfile.write(chunk)
```
maybe return a successfully log?

ywkim312 replied:
added print out for success

longshuicy commented on pyincore_data/utils/datautil.py:
```diff
@@ -51,3 +62,177 @@ def convert_dislocation_pd_to_csv(in_pd, save_columns, programname, savefile):
+            gpd.GeoDataFrame: GeoDataFrame with a new 'guid' column.
+        """
+        print("Creating GUID column")
+        for i, row in gdf.iterrows():
```
There might be faster way to add column without looping through each row. e.g.
```
gdf['guid'] = [str(uuid.uuid4()) for _ in range(len(gdf))]
```

longshuicy commented on pyincore_data/utils/datautil.py:
```diff
@@ -51,3 +62,177 @@ def convert_dislocation_pd_to_csv(in_pd, save_columns, programname, savefile):
+        statefips = fips[:2]
+        countyfips = fips[2:]
+        for i, row in gdf.iterrows():
+            guid_val = str(uuid.uuid4())
```
same as above there might be ways to avoid itterrows

longshuicy commented on pyincore_data/utils/datautil.py:
```diff
@@ -51,3 +62,177 @@ def convert_dislocation_pd_to_csv(in_pd, save_columns, programname, savefile):
+        return gdf
+
+    @staticmethod
+    def df_to_geopkg(gdf, outfile):
```
again my ocd...
gdf_to_geopkg?

longshuicy commented on pyincore_data/utils/datautil.py:
```diff
@@ -51,3 +62,177 @@ def convert_dislocation_pd_to_csv(in_pd, save_columns, programname, savefile):
+        gdf.to_file(outfile, driver="GPKG")
+
+    @staticmethod
+    def upload_postgres_from_gpk(infile):
```
upload_postgres_from_gpkg?

longshuicy reviewed (commented):
One thing I noticed is there are projections crs='EPSG:4326' coded in various functions. Wondering if we can put this into a variable in a central place and refer to that in the code.

This probably needs to be an independant issue after this get merged. At this moment, IN-CORE only uses 4326 but I was awaring of this issue and was thinking of adding the way of handling this but as you probably know, dealing with crs is much more complicated, only making it as variable might caused some other failure

ywkim312 commented on pyincore_data/nsiutil.py:
```diff
@@ -0,0 +1,152 @@
+            raise ValueError(f"State '{state_name}' not found. Please check the spelling.")
+
+        # Census API URL for county-level data
+        county_fips_url = f"https://api.census.gov/data/2020/acs/acs5?get=NAME&for=county:*&in=state:{state_fips}"
```
moved to globals

ywkim312 commented on pyincore_data/nsiutil.py:
```diff
@@ -0,0 +1,152 @@
+        # Extract county names and FIPS codes
+        county_list = [
+            {"county": row[0], "fips": f"{state_fips}{row[2]}"}
+            for row in counties_data[1:]  # Skip the header
```
yes, it should, otherwise, the error should be raised in previous lines

ywkim312 commented on pyincore_data/utils/datautil.py:
```diff
@@ -4,6 +4,17 @@
+
+from sqlalchemy import create_engine
+from geojson import FeatureCollection
+from ..config import Config as cfg
```
> relative import might be tricky. Could you add a `__init__.py` in the pyincore_data then see if you can directly import that package?

Changed

ywkim312 commented:
> One thing I noticed is there are projections crs='EPSG:4326' coded in various functions. Wondering if we can put this into a variable in a central place and refer to that in the code.


This probably needs to be an independant issue after this get merged. At this moment, IN-CORE only uses 4326 but I was awaring of this issue and was thinking of adding the way of handling this but as you probably know, dealing with crs is much more complicated, only making it as variable might caused some other failure

ywkim312 commented:
> The test folder does not need an `__init__.py` file. If you remove them from the "tests" folder and the "pyincore_data" folder everything should run fine. It looks like the top level one was added but not needed at some point and wasn't caught.

Got it so it is either to have the perfect init or just remove all in the tests. I updated the init.py in my previous push and it worked and this time I removed all of them in the tests folder and it worked. I will keep it as not having them in tests folder

longshuicy reviewed (approved):
pytests passed. Approve.
